### DISCUSSION TITLE
Фильтр каналов

### DISCUSSION THREAD

**Discussion by laika65** [melanchall/drywetmidi#169] [Q&A] [answered]
Существует ли в playback фильтр каналов? Если нет, то как лучше и проще реализовать его? Через EventCallback или собственным playback -ом?

```
  private static MidiEvent EventCallback(MidiEvent rawEvent, long rawTime, TimeSpan playbackTime)
  {
      //rawEvent можно ли тут получить канал? или обязательно надо приводить к типу?
      return null;
  }
```
И ещё нубский вопрос, каким тегом включать подсветку синтаксиса?

**Comment by melanchall** [comment, marked as answer]
Вы всё верно делаете, вот так можно с каналами логику организовать:

```csharp
private static MidiEvent EventCallback(MidiEvent rawEvent, long rawTime, TimeSpan playbackTime)
{
    var channelEvent = rawEvent as ChannelEvent;
    if (channelEvent == null)
        return rawEvent;

    // тут ваша логика по работе с channelEvent.Channel; чтобы проигнорировать событие, верните null
}
```

Подсветка синтаксиса делается через три бэктика (`) и сразу за ними слово csharp. Вот тут в первом же блоке показано: https://github.com/jgm/pandoc/issues/2117

**Reply by laika65** [reply]
Вот так и знал, что всё так просто. Огромное спасибо вам, продолжаю изучение библиотеки, и извиняюсь за такие простые вещи которые можно узнать из документации, но пока знаний языка не позволяют это делать.

**Reply by laika65** [reply]
Только сегодня начал делать фильтр, и ваш код всегда возвращает null. Почему то в rawEvent, вообще не прилетают ноты. И получается транспонировать ноты тоже не получиться, как в соседнем топике.

**Reply by melanchall** [reply]
Покажите свой код коллбэка полностью.

**Reply by laika65** [reply]
Фрагмент:
```csharp
private MidiEvent EventCallback(MidiEvent rawEvent, long rawTime, TimeSpan playbackTime)
{
    StringBuilder outputMessage = new StringBuilder();
    switch (rawEvent.EventType)
    {
        case MidiEventType.NoteOn:
            outputMessage.Append("NoteOn ").Append("Channel: ").Append(((NoteOnEvent)rawEvent).Channel.ToString());
            EventListBoxAdd(outputMessage);
            break;
        case MidiEventType.NoteOff:
            outputMessage.Append("NoteOff ").Append("Channel: ").Append(((NoteOnEvent)rawEvent).Channel.ToString());
            EventListBoxAdd(outputMessage);
            break;
        default:
            break;
    }
    return rawEvent;
}
```
ListBox не заполняется и с дебагом все эвенты кроме meta прилетают null
Миди файл:
[PIANOPIANO.zip](https://github.com/melanchall/drywetmidi/files/8332189/PIANOPIANO.zip)
Большинство файлов из миди библиотек такие, это не единичный случай.

**Reply by melanchall** [reply]
1. Да, события нот не попадают в [EventCallback](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_EventCallback), для нот есть [NoteCallback](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_NoteCallback) (по ссылке в документации и пример есть). Сделано это потому, что нота = пара событий. Поменяем мы что-то в NoteOn, нужно же ещё синхронизировать NoteOff. Поэтому был введён отдельный коллбэк, в котором можно обрабатывать ноты (не отдельные события). Там можете и транспонировать ноты и всё что угодно с ними делать, и быть уверенным, что пара событий будет меняться вместе.
2. Ваш конкретный код просто логгирует сообщения. Для этой цели используйте событие [Playback.EventPlayed](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_EventPlayed), оно вызывается для абсолютно всех MIDI-событий.

**Reply by melanchall** [reply]
И кстати, зачем так сложно составлять текст? Почему бы просто интерполяцию строк не использовать:

```csharp
EventListBoxAdd($"NoteOn Channel: {((NoteOnEvent)rawEvent).Channel}");
```

**Reply by laika65** [reply]
Прочитал, что надо делать через StringBuilder где то и делаю теперь везде, где конкатенация нужна. Ещё вопрос, куда прилетают Program Change, SysEx, Aftertouch и т.п.? В EventCallback? 
Сложная библиотека для начинающего, но спасибо что сделали, и продолжаете работать над ней. Планов много у меня по её использованию. Я сам музыкант, получил травму позвоночника, сейчас частично работают лишь руки. Начал писать программу для того, что бы продолжать заниматься любимым делом. Есть программа XotoPad, пользовался ей пол года, но она мне не очень подходить и маловато функций которые нужны мне. Спасибо вам за то, что отвечаете на довольно нубские вопросы.

**Reply by melanchall** [reply]
Большое уважение вам, что в такой ситуации (а тем более в такое непростое время) занимаетесь музыкой!

1. `StringBuilder` используется в "сложных" конкатенациях. Например, если вы собираете строку в разрозненных местах, или в цикле. Для вашего простого случая билдер не нужен, код становится необоснованно сложнее.
2. Все указанные вами события должны попадать в `EventCallback`. Если не попадают – это баг. Ну и ещё раз, `EventPlayed` будет вызвано для абсолютно всех событий. Если нужно просто логгировать данные, не изменяя их, используйте `EventPlayed`.
3. Понимаю, что библиотека сложная, но и сам протокол MIDI непрост, как и все сопутствующие ему процессы. Я в библиотеке стараюсь предоставить максимально гибкие возможности, так что сложность – вынужденные издержки :)

А что касется пользовательских программ вроде XotoPad: это именно готовые решения с интерфейсом. Кому-то этого достаточно, кому-то, как вам, хочется большего, каких-то кастомных решений. Тут без программирования не обойтись, конечно.
